p2p/wdrr_queue.go

```diff
@@ -0,0 +1,301 @@
+// assumed the flow's buffer is non-empty.
+func (f *drrFlow) pop() Envelope {
+	f.mtx.Lock()
+	defer f.mtx.Unlock()
+
+	e := f.buffer[0]
+	msgSize := uint(proto.Size(e.Message))
+
+	f.deficit -= msgSize
+	f.size -= msgSize
+	f.buffer = f.buffer[1:]
+
+	return e
+}
+
+// peek returns the Envelope at the HoQ, head of the queue. If the flow's buffer
+// is empty, an empty Envelope is returned.
+func (f *drrFlow) peek() Envelope {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+
+	if f.size == 0 || len(f.buffer) == 0 {
+		return Envelope{}
+	}
+
+	return f.buffer[0]
+}
+
+func (f *drrFlow) getDeficit() uint {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+
+	return f.deficit
+}
+
+func (f *drrFlow) getCapacity() uint {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+
+	return f.capacity
+}
+
+func (f *drrFlow) getSize() uint {
+	f.mtx.RLock()
+	defer f.mtx.RUnlock()
+
+	return f.size
+}
+
+// assert the WDDR queue implements the queue interface at compile-time
+var _ queue = (*wdrrQueue)(nil)
+
+// wdrrQueue implements a Weighted Deficit Round Robin, WDRR, scheduling
+// algorithm via the queue interface. A WDRR queue is created per peer, where
+// the queue will have N number of flows. Each flow corresponds to a p2p Channel,
+// so there are n input flows and a single output source, the peer's connection.
+//
+// Each flow has the following:
+// - fixed capacity: The number of bytes it can hold. Note, partial messages are
+//   not stored, so if a message exceeds the current size if the flow, even by a
+//   single byte, it will be dropped.
+// - size: The current number of bytes enqueued in the flow.
+// - quantum: The number of bytes that is added to the deficit counter of the
+//   flow in each round. The flow can send at most quantum bytes at a time. Each
+//   flow has its own unique quantum, which gives the queue its weighted nature.
+//   A higher quantum corresponds to a higher weight/priority.
+// - deficit counter: The number of bytes that the flow is allowed to transmit
+//   when it is its turn.
+//
+// See: https://en.wikipedia.org/wiki/Deficit_round_robin
+type wdrrQueue struct {
+	logger  log.Logger
+	peerID  NodeID
+	metrics *Metrics
+	chIDs   []ChannelID
+	flows   map[ChannelID]*drrFlow
+
+	closer        *tmsync.Closer
+	enqueueCh     chan Envelope
+	enqueueCloser *tmsync.Closer
+	dequeueCh     chan Envelope
+	dequeueCloser *tmsync.Closer
+}
+
+func newWDRRQueue(logger log.Logger, pID NodeID, m *Metrics, chDescs []ChannelDescriptor) *wdrrQueue {
+	// Copy the channel IDs from the channel descriptors and create a WDRR flow
+	// per ChannelDescriptor.
+	chIDs := make([]ChannelID, len(chDescs))
+	flows := make(map[ChannelID]*drrFlow)
+	for i, chDesc := range chDescs {
+		chID := ChannelID(chDesc.ID)
+		chIDs[i] = chID
+
+		capacity := chDesc.MaxSendCapacityBytes
+		if capacity == 0 {
+			capacity = chDesc.MaxSendBytes
+		}
+
+		flows[chID] = &drrFlow{
+			capacity: capacity,
+			quantum:  chDesc.MaxSendBytes,
```

Comment:
This seems wrong to me. From a quick glance, most channels don't appear to set `MaxSendBytes` at all using the default of 0, meaning the deficit is often never increased. And the channels that do set it often set it to several multiples of the maximum message size, which is a fairly large value. I'm not familiar with the specifics of WDRR, but I would imagine the quantum should be on the order of a typical message size, or even a fraction of it.

Reply:
That's because I do not have that code committed yet. I was experimenting with various `MaxSendBytes` and couldn't yield good results, hence this convo.